engine/table/src/main/java/io/deephaven/engine/table/impl/ZeroKeyUpdateByManager.java

```diff
@@ -0,0 +1,102 @@
+            @NotNull UpdateByOperator[] operators,
+            @NotNull UpdateByWindow[] windows,
+            @NotNull ColumnSource<?>[] inputSources,
+            @NotNull int[][] operatorInputSourceSlots,
+            @NotNull QueryTable source,
+            @NotNull final Map<String, ? extends ColumnSource<?>> resultSources,
+            @Nullable String timestampColumnName,
+            @NotNull UpdateByRedirectionHelper redirHelper,
+            @NotNull UpdateByControl control) {
+        super(source, operators, windows, inputSources, operatorInputSourceSlots, timestampColumnName, redirHelper,
+                control);
+
+        if (source.isRefreshing()) {
+            result = new QueryTable(source.getRowSet(), resultSources);
+
+            // this is a refreshing source, we will need a listener
+            listener = newUpdateByListener(description);
+            source.addUpdateListener(listener);
+            // result will depend on listener
+            result.addParentReference(listener);
+
+            // create input and output modified column sets
+            for (UpdateByOperator op : operators) {
+                op.createInputModifiedColumnSet(source);
+                op.createOutputModifiedColumnSet(result);
+            }
+
+            // create an updateby bucket instance directly from the source table
+            zeroKeyUpdateBy = new UpdateByBucketHelper(description, source, operators, windows, inputSources,
+                    operatorInputSourceSlots, resultSources, timestampColumnName, redirHelper, control);
+            buckets.offer(zeroKeyUpdateBy);
+
+            // make the source->result transformer
+            transformer = source.newModifiedColumnSetTransformer(result, source.getDefinition().getColumnNamesArray());
+
+            // result will depend on zeroKeyUpdateBy
+            result.addParentReference(zeroKeyUpdateBy);
+        } else {
+            zeroKeyUpdateBy = new UpdateByBucketHelper(description, source, operators, windows, inputSources,
+                    operatorInputSourceSlots, resultSources, timestampColumnName, redirHelper, control);
+            result = zeroKeyUpdateBy.result;
+            buckets.offer(zeroKeyUpdateBy);
+        }
+
+        // make a dummy update to generate the initial row keys
+        final TableUpdateImpl fakeUpdate = new TableUpdateImpl(source.getRowSet(),
+                RowSetFactory.empty(),
+                RowSetFactory.empty(),
+                RowSetShiftData.EMPTY,
+                ModifiedColumnSet.EMPTY);
+
+        // do the actual computations
+        final PhasedUpdateProcessor sm = new PhasedUpdateProcessor(fakeUpdate, true);
+        sm.processUpdate();
+    }
+
+    @Override
+    protected QueryTable result() {
+        return result;
+    }
+
+    @Override
+    protected boolean upstreamSatisfied(final long step) {
+        // for Zero-Key, only need to verify the source is satisfied
+        return source.satisfied(step);
```

Comment:
If you take my advice elsewhere and rely on the transformed PT's wrapped `Table` as a proxy for bucket satisfaction in the bucketed case, you will need to check the singular bucket helper's `result` for satisfaction here in addition to `source`.